minigrid/envs/wfc/wfcroom.py

```diff
@@ -0,0 +1,226 @@
+    "non_navigable",
+}
+
+EDGE_CONFIG = {
+    "navigable": EdgeDescriptor(between=("navigable",), structure="grid"),
+    "non_navigable": EdgeDescriptor(between=("non_navigable",), structure="grid"),
+    "start_goal": EdgeDescriptor(between=("start", "goal"), structure=None),
+    # "lava_goal": EdgeDescriptor(between=("lava", "goal"), weight="lava_prob"),
+    # "moss_goal": EdgeDescriptor(between=("moss", "goal"), weight="moss_prob"),
+}
+
+
+class WFCEnv(MiniGridEnv):
+    """
+    ## Description
+
+    This environment procedurally generates a level using the Wave Function Collapse algorithm.
+    The environment supports a variety of different level structures but the default is a simple maze.
+    Requires the optional dependencies `imageio` and `networkx` to be installed with `pip install minigrid[wfc]`.
+
+    ## Mission Space
+
+    "traverse the maze to get to the goal"
+
+    ## Action Space
+
+    | Num | Name         | Action                    |
+    |-----|--------------|---------------------------|
+    | 0   | left         | Turn left                 |
+    | 1   | right        | Turn right                |
+    | 2   | forward      | Move forward              |
+    | 3   | pickup       | Unused                    |
+    | 4   | drop         | Unused                    |
+    | 5   | toggle       | Unused                    |
+    | 6   | done         | Unused                    |
+
+    ## Observation Encoding
+
+    - Each tile is encoded as a 3 dimensional tuple:
+        `(OBJECT_IDX, COLOR_IDX, STATE)`
+    - `OBJECT_TO_IDX` and `COLOR_TO_IDX` mapping can be found in
+        [minigrid/minigrid.py](minigrid/minigrid.py)
+    - `STATE` refers to the door state with 0=open, 1=closed and 2=locked
+
+    ## Rewards
+
+    A reward of '1 - 0.9 * (step_count / max_steps)' is given for success, and '0' for failure.
+
+    ## Termination
+
+    The episode ends if any one of the following conditions is met:
+
+    1. The agent reaches the goal.
+    2. Timeout (see `max_steps`).
+
+    ## Registered Configurations
+
+    S: size of map SxS.
+
+    """
+
+    PATTERN_COLOR_CONFIG = {
+        "wall": (0, 0, 0),  # black
+        "empty": (255, 255, 255),  # white
+    }
+
+    def __init__(
+        self,
+        wfc_config: WFCConfig | str = "MazeSimple",
+        size: int = 25,
+        ensure_connected: bool = True,
+        max_steps: int | None = None,
+        **kwargs,
+    ):
+        self.config = (
+            wfc_config if isinstance(wfc_config, WFCConfig) else WFC_PRESETS[wfc_config]
+        )
+        self.padding = 1
+
+        # This controls whether to process the level such that there is only a single connected navigable area
+        self.ensure_connected = ensure_connected
+
+        mission_space = MissionSpace(mission_func=self._gen_mission)
+
+        if size < 3:
+            raise ValueError(f"Grid size must be at least 3 (currently {size})")
+        self.size = size
+        self.max_attempts = 100
```

Comment:
Best to set default value to math.inf (as in MultiRoomEnv), and possibly setting max_attemps as an init parameter.